### DISCUSSION TITLE
How to attach an image to an existing artifact

### DISCUSSION THREAD

**Discussion by usrme** [oras-project/oras#1560] [Q&A]
Right off the bat: I'm terribly sorry if I'm using the wrong terminology throughout; I'm still not well-versed in the space of OCI!

What I'm basically trying to achieve is to have an artifact that just refers to other images. This artifact would serve as a blueprint for a given environment in terms of what images that environment requires and if a new environment receives the same blueprint, then the list of references could be iterated to, for example, pull all the images.

I've tried to find examples of other people doing something similar, but I've only mostly found people attaching annotations to images, whereas I'd need a way to attach an image to an artifact.

Something akin to the following doesn't seem to work:

```console
$ oras attach --artifact-type application/vnd.oci.image.config.v1+json localhost:5000/my-artifact:v1 localhost:5000/hello-world:latest
✓ Exists    application/vnd.oci.empty.v1+json
  └─ sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
✓ Uploaded  localhost:5000/hello-world
  └─ sha256:0e48cc301f3dcc37cf3e4bdda58bd25dfa7c5cbcd6289e201efd29f132f085aa
⠋ [                    ]( 128 kB/s) Uploading application/vnd.oci.image.manifest.v1+json
  └─ sha256:6a525fb29ae08b5504a9966cb2a07ec25ef9ca9c6f4293310ef84502a1723b36
Error response from registry: manifest invalid: manifest invalid: map[description:
During upload, manifests undergo several checks ensuring validity. If those checks fail, 
this error MAY be returned, unless a more specific error is included. The detail will 
contain information the failed validation. jsonSchemaValidation:[layers.0.mediaType: Does 
not match pattern '^[A-Za-z0-9][A-Za-z0-9!#$&-^_.+]{0,126}/[A-Za-z0-9][A-Za-z0-9!#$&-^_.+]{0,126}$'] 
reference:sha256:6a525fb29ae08b5504a9966cb2a07ec25ef9ca9c6f4293310ef84502a1723b36]
```

Where `localhost:5000/hello-world:latest` is created using `oras cp docker.io/library/hello-world:latest --to-oci-layout localhost:5000/hello-world:latest --platform linux/amd64` and the custom artifact is created using:

```console
$ echo "{\"name\":\"foo\",\"value\":\"bar\"}" > config.json
$ oras push localhost:5000/my-artifact:v1 --config config.json:application/vnd.com.company.config.v1+json
```

Any help is much appreciated!

**Comment by qweeah** [comment]
@usrme I think you have misused the command. The last argument, `localhost:5000/hello-world:latest`, means you want to upload a local file `localhost:5000/hello-world` with media type as `latest`. `latest` fails the registry (running at localhost:5000) check (regex match of `^[A-Za-z0-9][A-Za-z0-9!#$&-^_.+]{0,126}/[A-Za-z0-9][A-Za-z0-9!#$&-^_.+]{0,126}$`).

**Reply by usrme** [reply]
Would you be able to let me know whether and how it would be possible to attach images to an artifact? I'm still misunderstanding the process, it seems.
